src/qibolab/instruments/emulator/pulse_simulator.py

```diff
@@ -0,0 +1,711 @@
+        return results
+
+    def run_pulse_simulation(
+        self,
+        sequence: PulseSequence,
+        instant_measurement: bool = True,
+    ) -> tuple[np.ndarray, list]:
+        """Simulates the input pulse sequence.
+
+        Args:
+            sequence (`qibolab.pulses.PulseSequence`): Pulse sequece to simulate.
+            instant_measurement (bool): Collapses readout pulses to duration of 1 if True. Defaults to True.
+
+        Returns:
+            tuple: Reduced density matrix of final state specified by readout qubits as well as the list of readout qubits in, both in little endian order.
+        """
+        # reduces measurement time to 1 dt to save simulation time
+        if instant_measurement:
+            for i in range(len(sequence)):
+                if type(sequence[i]) is ReadoutPulse:
+                    sequence[i].duration = 1
+
+        # extract waveforms from pulse sequence
+        channel_waveforms = ps_to_waveform_dict(
+            sequence,
+            self.platform2simulator_channels,
+            self.sampling_rate,
+            self.sim_sampling_boost,
+            self.runcard_duration_in_dt_units,
+        )
+
+        self.pulse_sequence_history.append(sequence.copy())
+        self.channel_waveforms_history.append(channel_waveforms)
+        # execute pulse simulation in emulator
+        ro_reduced_dm, rdm_qubit_list = self.simulation_backend.qevolve(
+            channel_waveforms, self.simulate_dissipation
+        )
+
+        return ro_reduced_dm, rdm_qubit_list
+
+    def play(
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers: Dict[QubitId, Coupler],
+        sequence: PulseSequence,
+        execution_parameters: Optional[ExecutionParameters] = None,
+    ) -> dict[str, Union[IntegratedResults, SampleResults]]:
+        """Executes the sequence of instructions and generates readout results.
+
+        Args:
+            qubits (dict): Qubits involved in the device. Does not affect emulator.
+            couplers (dict): Couplers involved in the device. Does not affect emulator.
+            sequence (`qibolab.pulses.PulseSequence`): Pulse sequence to simulate.
+            execution_parameters (`qibolab.ExecutionParameters`): Parameters (nshots,
+                                                        relaxation_time,
+                                                        fast_reset,
+                                                        acquisition_type,
+                                                        averaging_mode)
+                                                        Defaults to None, for which case it uses the values stored in self.exec_params.
+
+        Returns:
+            dict: A dictionary mapping the readout pulses serial and respective qubits to
+            Qibolab results object.
+        """
+        if execution_parameters is None:
+            execution_parameters = self.exec_params
```

Comment:
Following a comment from above, I would remove this
```suggestion
```
and make `execution_parameters` a required attribute, since the `Platform` is always passing it anyway: https://github.com/qiboteam/qibolab/blob/fcdd2af2a28f85995aae1370b0bb095324a4926f/src/qibolab/platform.py#L171

Reply:
Fixed in 170b78df1260f4f8dad1b400c63a0fa59a3227f0